common/lib/xmodule/xmodule/util/adaptive_learning.py

```diff
@@ -0,0 +1,270 @@
+        """
+        Return 'knowledge node student' object for user identified by `user_id`
+        and unit identified by `block_id`.
+        """
+        # Create student
+        self.get_or_create_student(user_id)
+        # Link student to unit
+        knowledge_node_student = self.get_knowledge_node_student(block_id, user_id)
+        if knowledge_node_student is None:
+            knowledge_node_student = self.create_knowledge_node_student(block_id, user_id)
+        return knowledge_node_student
+
+    def get_or_create_student(self, user_id):
+        """
+        Create a new student on external service if it doesn't exist,
+        and return it.
+        """
+        student = self.get_student(user_id)
+        if student is None:
+            student = self.create_student(user_id)
+        return student
+
+    def get_student(self, user_id):
+        """
+        Return external information about student identified by `user_id`,
+        or None if external service does not know about student.
+        """
+        students = self.get_students()
+        try:
+            student = next(s for s in students if s.get('uid') == user_id)
+        except StopIteration:
+            student = None
+        return student
+
+    def get_students(self):
+        """
+        Return list of all students that external service knows about.
+        """
+        url = self.students_url
+        response = requests.get(url, headers=self.request_headers)
+        students = json.loads(response.content)
+        return students
+
+    def create_student(self, user_id):
+        """
+        Create student identified by `user_id` on external service,
+        and return it.
+        """
+        url = self.students_url
+        payload = {'uid': user_id}
+        response = requests.post(url, headers=self.request_headers, data=payload)
+        student = json.loads(response.content)
+        return student
+
+    def get_knowledge_node_student(self, block_id, user_id):
+        """
+        Return 'knowledge node student' object for user identified by `user_id`
+        and unit identified by `block_id`, or None if it does not exist.
+        """
+        # Get 'knowledge node student' objects
+        links = self.get_knowledge_node_students()
+        # Filter them by `block_id` and `user_id`
+        try:
+            link = next(
+                l for l in links if l.get('knowledge_node_uid') == block_id and l.get('student_uid') == user_id
+            )
+        except StopIteration:
+            link = None
+        return link
+
+    def get_knowledge_node_students(self):
+        """
+        Return list of all 'knowledge node student' objects for this course.
+        """
+        url = self.knowledge_node_students_url
+        response = requests.get(url, headers=self.request_headers)
+        links = json.loads(response.content)
+        return links
+
+    def create_knowledge_node_student(self, block_id, user_id):
```

Comment:
@pomegranited Yup, that's a great idea. Done in a8ccc570d444a04520e71a8c1bff2959f65846cc.